### ISSUE TITLE
C++ tutorials?

### ISSUE THREAD

**Issue by charlesrwest** [facebookarchive/caffe2#351]
Hello!  First off, I am really impressed by the work that has been done so far and like the changes from caffe 1.

If I may ask, are there any plans to have any tutorials on how to do basic operations in C++ or a section/split in the website for python vs c++ tutorials?  Caffe 1 has a C++ API, but it is completely undocumented and there are not really many tutorials for it (https://medium.com/@shiyan/caffe-c-helloworld-example-with-memorydata-input-20c692a82a22).

I am fairly familiar with protobuf and can parse the workflow to hack things together, but it takes a rather large amount of digging through the source code to determine how to do relatively basic things (which are covered in python tutorials).  I understand that Python is the dominate language in machine learning, but it seems odd to more or less completely ignore one of the other most popular ones (C/C++ in 2nd/3rd place according to some lists: https://www.techworm.net/2017/03/top-20-popular-programming-languages-2017.html).

I would be happy to help if someone is willing to show me where to post and double check to make sure that I am not leading the user base horribly astray.

**Comment by Yangqing** [comment]
This is definitely a great idea. Indeed we are short on C++ tutorials right now, while in a lot of products we are using C++ directly instead of Python for apparent reasons. Will start a project and do more C++ examples - something like nvidia's CUDA samples.

**Comment by bcaine** [comment]
Awesome @Yangqing. I haven't dug completely through the source yet, but I didn't see this on the website. Is there a simple forward pass C++ example somewhere? 

Basically just loading a network with pretrained weights, passing data (say images) into memory, and pulling out the result from the last layer?

Thanks!

**Comment by Yangqing** [comment]
Ah, for that the predictor verifier code might be interesting:

https://github.com/caffe2/caffe2/blob/master/caffe2/binaries/predictor_verifier.cc

and also the predictor test that shows how we do I/Os into and out of the predictor:

https://github.com/caffe2/caffe2/blob/5bd6b21ddfd9c5cc87c528dab1c78e7ab0f49259/caffe2/core/predictor_test.cc

**Comment by mfojtak** [comment]
+1 for this enhancement

It would also be great to have cling based C++ jupyter notebooks with tutorials. I have no problem creating them but there is no documentation on how to use Caffe2 from C++ in the first place.

**Comment by crohkohl** [comment]
:+1: 

What I always loved about caffe was the "C++ first" style of development. Maybe I am wrong, but for me personally it is really sad to see that it has changed into yet another python driven framework.  It would be great to have an example how to train a network using plain old simple C++ without going through all the python hassle. 

However, I really like the new library design. Good work!

**Comment by charlesrwest** [comment]
Agreed.  I'm having some build issues with the library on my GPU machine, but I intend to try to reverse engineer the MNIST example.  It appears most of what is called has a related C++ class, with the exception of the LENET convenience wrapper.  Looking at the resulting prototext from the python example should make it possible to make a functional example without that.

**Comment by ThatAIGeek** [comment]
When they are available I might also help to create Cling notebooks, so looking forward to them

**Comment by jasjuang** [comment]
+1 for this

It would so nice to have a C++ tutorial

**Comment by Soledad89** [comment]
Great, C++ tutorial will help me a lot.

**Comment by charlesrwest** [comment]
I did a little work on replicating the MNIST tutorial in C++.  The main problem I am running into is that there is a set of Python helper functions used to define the convolutional network architectures which are not replicated in C++.  If anyone is able to convert these functions, it would go a long way toward having side by side tutorials.

In the mean time, I have determined (and commented) how to load and do some training on the network architectures that have been previously created (as pbtxt by those same python functions).

I've attached the source code and associated network files.  Please tell me what you think.


  
[tutorialCodeAndNetworkDefinitions.zip](https://github.com/caffe2/caffe2/files/1098419/tutorialCodeAndNetworkDefinitions.zip)

**Comment by charlesrwest** [comment]
`#include<iostream>
#include<string>
#include "caffe2/core/init.h"
#include "caffe2/core/operator.h"
#include "caffe2/proto/caffe2.pb.h"
#include "caffe2/utils/proto_utils.h"
#include "caffe2/core/logging.h"

#include "caffe2/core/workspace.h"



//Can store/manage blobs and nets in Workspace class, can also be used to run a plan

/**
This program expects two directory paths to be passed in:
Arg1: directory containing folders mnist-train-nchw-leveldb and mnist-test-nchw-leveldb
Arg2: directory containing network definitions to use
Arg3: output directory for generated files
*/
int main(int argc, char **argv)
{

//Get folders to access resources
if(argc < 4)
{
std::cerr << "Insufficient arguments:" << std::endl;
std::cerr << "Required:" << std::endl;
std::cerr << "Directory path containing folders mnist-train-nchw-leveldb and mnist-test-nchw-leveldb" << std::endl;
std::cerr << "Directory path containing network definitions (.pbtxt)" << std::endl;
std::cerr << "Directory path to place generated files" << std::endl;
return 1;
}

std::string training_data_folder(argv[1]);
std::string networkDefinitionFolder(argv[2]);
std::string output_folder(argv[3]);

//Initialize workspace which all data/calculation blobs will reside in
caffe2::Workspace workspace(output_folder);

//Initialization networks don't actually do any computations.  They just define/initialize the blobs that will be used in the other network definitions (shared and living in the workspace).

//This is why they are created and then run once before the other networks are.  The training initialization network fills all of the weights with their initialization values, while the test initialization network just loads the databases to test with without touching the weights

//Retrieve the protobuf object which defines the initialization values for the network from a text file
caffe2::NetDef trainingInitializationNetworkDefinition;

if(!ReadProtoFromFile(networkDefinitionFolder + "/train_init_net.pbtxt", &trainingInitializationNetworkDefinition))
{
std::cerr << "Error opening initialization network definition" << std::endl;
return 1;
}

//Load the protobuf object with training network architecture definition from a text file.  This network references the same blobs in the workspace that are set with the initialization network.
caffe2::NetDef  trainingNetworkDefinition;

if(!ReadProtoFromFile(networkDefinitionFolder + "/train_net.pbtxt", &trainingNetworkDefinition))
{
std::cerr << "Error opening training network definition" << std::endl;
return 1;
}

//Run the initialization network so that the blobs that the training network relies on are loaded and set with the initial values.
if(!workspace.RunNetOnce(trainingInitializationNetworkDefinition))
{
std::cerr << "Error loading/running training network initialization" << std::endl;
return 1;
}

//Define the training network in the workspace based on the training network protobuf object
caffe2::NetBase* trainingNetworkPointer = workspace.CreateNet(trainingNetworkDefinition);
if(trainingNetworkPointer == nullptr)
{
std::cerr << "Error instancing training network" << std::endl;
return 1;
}

//Train the network for 200 interations and report the accuracy at each step
int32_t numberOfTrainingRuns = 200;

for(int32_t runIndex = 0; runIndex < numberOfTrainingRuns; runIndex++)
{
//Run an iteration of the training network
workspace.RunNet(trainingNetworkDefinition.name());

//Retrieve the reported "accuracy" so that it can be printed out (1x1)
const caffe2::Blob* accuracyBlobPointer = workspace.GetBlob("accuracy");
if(accuracyBlobPointer == nullptr)
{
std::cerr << "Error, unable to access training accuracy blob" << std::endl;
return 1;
}

std::cout << "Accuracy: " <<  *(accuracyBlobPointer->Get<caffe2::Tensor<caffe2::CPUContext>>().data<float>()) << std::endl;

//Retrieve the reported loss so that it can be printed out (1x1)
const caffe2::Blob* lossBlobPointer = workspace.GetBlob("loss");
if(lossBlobPointer == nullptr)
{
std::cerr << "Error, unable to access training loss blob" << std::endl;
return 1;
}

std::cout << "Loss: " <<  *(lossBlobPointer->Get<caffe2::Tensor<caffe2::CPUContext>>().data<float>()) << std::endl;

//Both accuracy and loss are tensors
}

//Report success
std::cout << "All operations completed without error" << std::endl;

return 0;
} `

**Comment by leovandriel** [comment]
This might get you up to speed:

https://github.com/leonardvandriel/caffe2_cpp_tutorial

It covers the Intro, Toy, Squeezenet and MNIST tutorials. Let me know if there's any particular tutorial you're interested in getting done as well.

**Comment by charlesrwest** [comment]
Huzzah!  Truly you are a saint among men.  Do you think there is any way we could get this on the Caffe2 site?

Seriously, nice work.

**Comment by leovandriel** [comment]
I would be fine with it moving there.

At this point I find the Caffe2 docs and tutorials fairly messy and incomplete, so perhaps it good to have this grow on the side for a little bit.

**Comment by charlesrwest** [comment]
I've been pondering the best way to make helper classes in c++ similar to what there are in Python.  

What I eventually settled on is a set of virtual classes based on a "ComputeModuleDefinition".  At the most basic level, a compute module can stand in for an operator and switch out what is needed depending on the network mode (such as "TRAIN", "TEST" or "DEPLOY").  It also provides one or more initialization operators on demand.  A good example is "SoftMax" and "SoftMaxWithLoss".  In general, you only want the loss part when you are doing training or testing.  Defining a compute module allows which operator ends up in the network to be dynamic on the intention for the network.

A second important piece of the concept is a virtual class derived from a ComputeModuleDefinition that allows multiple compute modules to be grouped together to for a single module (with most of the boiler plate being automatically generated via inheritance).  This allows you to define complex building blocks, such a resnet module or a GoogLenet section without having to do a ton of copy and pasting.  In my example code, I define a module which allows you to make fully connected networks of arbitrary depth and width as a single module.

Defining a few virtual functions also allows you to do some nice automatic generation, such as automatically generating all of the ADAM operators needed to do training for a network and gradients.

Please let me know what you guys think.  I believe this might be useful for working with Caffe2 in C++, but I would be interested to know what others are doing.

Links:
Learning a Sine wave using modules:
https://github.com/charlesrwest/Caffe2ComputeModules/blob/master/src/executables/trainSineWave/main.cpp

Compute Module Definition:
https://github.com/charlesrwest/Caffe2ComputeModules/blob/master/src/library/ComputeModuleDefinition.hpp

Composite Compute Module Definition:
https://github.com/charlesrwest/Caffe2ComputeModules/blob/master/src/library/CompositeComputeModuleDefinition.hpp

**Comment by jwatte** [comment]
@leonardvandriel What is the "NetUtil" class your MNIST C++ example uses? I can't find it in the Caffe2 repo.

**Comment by leovandriel** [comment]
It's part of the caffe2_cpp_tutorial repo, see [net.h](https://github.com/leonardvandriel/caffe2_cpp_tutorial/blob/master/include/caffe2/util/net.h) and [net.cc](https://github.com/leonardvandriel/caffe2_cpp_tutorial/blob/master/src/caffe2/util/net.cc). I put it in the same namespace and similar include path, which is maybe a bit confusing. It's like an extension to the Caffe2 repo.

**Comment by ahkarami** [comment]
Really Caffe2 requires a nice C++ tutorial (like PyTorch which has some great ones in python).

**Comment by Crefeda** [comment]
Hi,

Is there any tutorial on how to use speed_benchmark.cc or caffe2_benchmark.cc. I am specifically looking to pass input images, convert it into the format for the model to ingest and output a class probability. Any pointers would be of great help.
